crates/bevy_ecs/src/query/builder.rs

```diff
@@ -81,15 +81,8 @@ impl<'w, D: QueryData, F: QueryFilter> QueryBuilder<'w, D, F> {
                 .is_some_and(|info| info.storage_type() == StorageType::Table)
         };
 
-        let (mut component_reads_and_writes, component_reads_and_writes_inverted) =
-            self.access.access().component_reads_and_writes();
-        if component_reads_and_writes_inverted {
-            return false;
-        }
-
-        component_reads_and_writes.all(is_dense)
-            && self.access.access().archetypal().all(is_dense)
-            && !self.access.access().has_read_all_components()
+        D::IS_DENSE
```

Comment:
I think we might want more comments here to explain why this is ok

Reply:
Good idea!  I added some comments here, and some in `QueryState::transmute_filtered()`.  (I thought to write up an explanation in the PR description, but I didn't think to put it in the code...)

Reply:
I might even add that `FilteredEntityRef|Mut` (which is one of the primary use-cases of QueryBuilder) has D::IS_DENSE=True, so adding `Option<&Sparse>` will maintain dense-ness.

(IIUC, the key point being, as i wrote in the other comment, that dense-ness only depends on required components (because for those components we need to iterate only through the archetypes that have that component, and not through all tables)?
And for optional components we can just always just iterate through the tables.)

Should the check consider all `required` components on top of filters?
Oh I see, it's what you were mentioning, for "normal" query data like `&, Ref, &mut ,Mut` `D::IS_DENSE` is enough (since those query data add the access to required)